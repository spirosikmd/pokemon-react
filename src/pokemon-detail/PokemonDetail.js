import React, { PureComponent } from 'react';
import { Modal, Image, List, Step, Grid } from 'semantic-ui-react';

class PokemonDetail extends PureComponent {
  getDescriptionEvolutionGroupItem(evolutionRequirements) {
    return `${evolutionRequirements.amount} ${evolutionRequirements.name}`;
  }

  getOwnEvolutionGroupItem({ pokemon }) {
    return {
      title: pokemon.name,
      active: true,
      description: this.getDescriptionEvolutionGroupItem(
        pokemon.evolutionRequirements
      ),
    };
  }

  getEvolutionsStepGroupItems(evolutions) {
    let evolutionGroupItems = evolutions.map(evolution => {
      let data = { title: evolution.name };
      if (evolution.evolutionRequirements) {
        const description = this.getDescriptionEvolutionGroupItem(
          evolution.evolutionRequirements
        );
        data = Object.assign({}, { description }, data);
      }
      return data;
    });
    evolutionGroupItems = [
      this.getOwnEvolutionGroupItem(this.props),
      ...evolutionGroupItems,
    ];
    return evolutionGroupItems;
  }

  render() {
    const pokemon = this.props.pokemon;

    return (
      <Modal open={this.props.show} onClose={this.props.onCloseModal}>
        <Modal.Header>{pokemon.name}</Modal.Header>
        <Modal.Content image>
          <Image src={pokemon.image} width="190" height="160" />
          <Modal.Description>
            {pokemon.evolutions &&
              <Step.Group
                size="mini"
                items={this.getEvolutionsStepGroupItems(pokemon.evolutions)}
              />}
            <Grid columns={2} container>
              <Grid.Column>
                <List>
                  <List.Item>
                    <List.Header>Classification</List.Header>
                    {pokemon.classification}
                  </List.Item>
                  <List.Item>
                    <List.Header>Types</List.Header>
                    {pokemon.types.join(', ')}
                  </List.Item>
                  <List.Item>
                    <List.Header>Max HP</List.Header>
                    {pokemon.maxHP}
                  </List.Item>
                  <List.Item>
                    <List.Header>Max CP</List.Header>
                    {pokemon.maxCP}
                  </List.Item>
                  <List.Item>
                    <List.Header>Weight</List.Header>
                    {pokemon.weight.minimum} - {pokemon.weight.maximum}
                  </List.Item>
                  <List.Item>
                    <List.Header>Height</List.Header>
                    {pokemon.height.minimum} - {pokemon.height.maximum}
                  </List.Item>
                  {pokemon.habitat &&
                    <List.Item>
                      <List.Header>Habitat</List.Header>
                      {pokemon.habitat}
                    </List.Item>}
                </List>
              </Grid.Column>
              <Grid.Column>
                <List>
                  <List.Item>
                    <List.Header>Resistant</List.Header>
                    {pokemon.resistant.join(', ')}
                  </List.Item>
                  <List.Item>
                    <List.Header>Weaknesses</List.Header>
                    {pokemon.weaknesses.join(', ')}
                  </List.Item>
                  <List.Item>
                    <List.Header>Flee rate</List.Header>
                    {pokemon.fleeRate}
                  </List.Item>
                  <List.Item>
                    <List.Header>Attacks</List.Header>
                    <ul>
                      {pokemon.attacks.fast.map(fastAttack => (
                        <li key={fastAttack.name}>
                          {fastAttack.name}
                          {' - '}
                          {fastAttack.type}
                          {' - '}
                          {fastAttack.damage}
                        </li>
                      ))}
                    </ul>
                    <ul>
                      {pokemon.attacks.special.map(specialAttack => (
                        <li key={specialAttack.name}>
                          {specialAttack.name}
                          {' - '}
                          {specialAttack.type}
                          {' - '}
                          {specialAttack.damage}
                        </li>
                      ))}
                    </ul>
                  </List.Item>
                </List>
              </Grid.Column>
            </Grid>
          </Modal.Description>
        </Modal.Content>
      </Modal>
    );
  }
}

export default PokemonDetail;
