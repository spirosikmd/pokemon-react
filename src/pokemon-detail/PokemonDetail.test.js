import React from 'react';
import renderer from 'react-test-renderer';
import PokemonDetail from './PokemonDetail';

const pokemon = {
  name: 'pikachu',
  weight: 100,
  height: 100,
  base_experience: 100,
  order: 20,
  species: {
    name: 'pikachu',
  },
  abilities: [
    {
      ability: {
        name: 'thunder',
      },
    },
  ],
  forms: [
    {
      name: 'pikachu',
    },
  ],
};

test('PokemonDetail renders pokemon details', () => {
  const component = renderer.create(
    <PokemonDetail pokemon={pokemon} show={true} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
