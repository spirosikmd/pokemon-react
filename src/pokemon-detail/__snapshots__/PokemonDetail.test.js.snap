exports[`test PokemonDetail renders pokemon details 1`] = `<noscript />`;
